[mlir] Add memref normalization support for reinterpret_cast op

arnab-polymage opened the pull request:
Rewrites the memref defined by reinterpet_cast op to have an identity layout map
and updates all its indexing uses. Also extend `replaceAllMemRefUsesWith` utility
to work when there are multiple occurences of `oldMemRef` in `op`'s operand list
when op is non-dereferencing.

Fixes #122090 
Fixes #121091

bondhugula commented on mlir/lib/Dialect/Affine/Utils/Utils.cpp:
```diff
@@ -1846,6 +1785,94 @@ LogicalResult mlir::affine::normalizeMemRef(AllocLikeOp *allocOp) {
+  ArrayRef<int64_t> newShape = newMemRefType.getShape();
+  // Compute size along all the dimensions of the new normalized memref.
+  for (unsigned i = 0; i < newRank; i++) {
+    if (newShape[i] != ShapedType::kDynamic)
```
Use `isDynamicDim`.

bondhugula commented on mlir/include/mlir/Dialect/Affine/Utils.h:
```diff
@@ -243,16 +244,18 @@ LogicalResult replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
 extern template LogicalResult
 normalizeMemRef<memref::AllocOp>(memref::AllocOp *op);
 
+LogicalResult normalizeMemRef(memref::ReinterpretCastOp *op);
```
Missing doc comment. Just refer to the previous one or drop the blank line above.

bondhugula commented on mlir/include/mlir/Dialect/Affine/Utils.h:
```diff
@@ -243,16 +244,18 @@ LogicalResult replaceAllMemRefUsesWith(Value oldMemRef, Value newMemRef,
+/// Rewrites the memref defined by alloc or reinterpret_cast op to have an
+/// identity layout map and updates all its indexing uses. Returns failure if
+/// any of its uses escape (while leaving the IR in a valid state).
 template <typename AllocLikeOp>
 LogicalResult normalizeMemRef(AllocLikeOp *op);
 extern template LogicalResult
 normalizeMemRef<memref::AllocaOp>(memref::AllocaOp *op);
 extern template LogicalResult
 normalizeMemRef<memref::AllocOp>(memref::AllocOp *op);
```
Doc comment is missing anything on why a pointer to the derived op type is being passed.

arnab-polymage replied:
It is not clear to me as well. I just followed the exisiting convention (used pointer for reinterpret_cast op as we alrady had pointers for alloc ops).

bondhugula replied:
But it is an input argument or output argument or both? Is it nullable or not? It's time to update such a key thing in the doc comment.

bondhugula replied:
I've fixed that here: https://github.com/llvm/llvm-project/pull/134466

bondhugula commented on mlir/lib/Dialect/Affine/Utils/Utils.cpp:
```diff
@@ -1098,90 +1098,12 @@ void mlir::affine::affineScalarReplace(func::FuncOp f, DominanceInfo &domInfo,
-
-  return success();
+// Checks if `op` is non dereferencing.
+// TODO: This hardcoded check will be removed once the right interface is added.
```
Should this TODO be removed now?

arnab-polymage replied:
No

bondhugula commented on mlir/lib/Dialect/Affine/Utils/Utils.cpp:
```diff
@@ -1216,53 +1138,57 @@ LogicalResult mlir::affine::replaceAllMemRefUsesWith(
   OpBuilder builder(op);
   // The following checks if op is dereferencing memref and performs the access
   // index rewrites.
   auto affMapAccInterface = dyn_cast<AffineMapAccessInterface>(op);
```
If `affMap...` isn't used here any more, we shouldn't be dyn_casting it here. Move to its first use site.

bondhugula commented on mlir/lib/Dialect/MemRef/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -353,16 +354,25 @@ void NormalizeMemRefs::normalizeFuncOpMemRefs(func::FuncOp funcOp,
   // Turn memrefs' non-identity layouts maps into ones with identity. Collect
   // alloc/alloca ops first and then process since normalizeMemRef
   // replaces/erases ops during memref rewriting.
```
This comment is out of date now.

bondhugula commented on mlir/lib/Dialect/MemRef/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -353,16 +354,25 @@ void NormalizeMemRefs::normalizeFuncOpMemRefs(func::FuncOp funcOp,
-  funcOp.walk([&](memref::AllocaOp op) { allocaOps.push_back(op); });
-  for (memref::AllocaOp allocaOp : allocaOps)
+  SmallVector<AllocaOp> allocaOps;
+  funcOp.walk([&](AllocaOp op) { allocaOps.push_back(op); });
+  for (AllocaOp allocaOp : allocaOps)
     (void)normalizeMemRef(allocaOp);
 
+  // Turn memrefs' non-identity layouts maps into ones with identity. Collect
+  // reinterpret_cast ops first and then process since normalizeMemRef
+  // replaces/erases ops during memref rewriting.
+  SmallVector<ReinterpretCastOp> reinterpretCastOps;
+  funcOp.walk(
+      [&](ReinterpretCastOp op) { reinterpretCastOps.push_back(op); });
```
Instead of walking the function thrice, can we walk once to collect all, or does the processing of one impact the other lists?

arnab-polymage replied:
It won't currently becaue we process only dereferencing uses of an op which is to be normalized.

bondhugula commented:
Fix format as well.

bondhugula reviewed (commented):
Looking mostly good. Some comments.

9Tempest commented on mlir/lib/Dialect/Affine/Utils/Utils.cpp:
```diff
@@ -1845,6 +1781,95 @@ LogicalResult mlir::affine::normalizeMemRef(AllocLikeOp allocOp) {
+  // Fetch a new memref type after normalizing the old memref to have an
+  // identity map layout.
+  MemRefType newMemRefType = normalizeMemRefType(memrefType);
+  newMemRefType.dump();
```
Why do a `dump` here? Is this for debugging purpose?

arnab-polymage replied:
Thanks for catching this. Will remove.

arnab-polymage commented on mlir/lib/Dialect/Affine/Utils/Utils.cpp:
```diff
@@ -1216,53 +1138,55 @@ LogicalResult mlir::affine::replaceAllMemRefUsesWith(
+  AffineMap oldMap;
+  unsigned oldMemRefNumIndices = oldMemRefRank;
+  auto startIdx = op->operand_begin() + memRefOperandPos + 1;
+  auto affMapAccInterface = dyn_cast<AffineMapAccessInterface>(op);
+  if (affMapAccInterface) {
```
No point doing that as it is used in line 1188 below as well.

arnab-polymage commented on mlir/test/Dialect/MemRef/normalize-memrefs.mlir:
```diff
@@ -3,8 +3,8 @@
 
-// CHECK-DAG: #[[$REDUCE_MAP1:.*]] = affine_map<(d0, d1) -> ((d0 mod 2) * 2 + d1 mod 2 + (d0 floordiv 2) * 4 + (d1 floordiv 2) * 8)>
-// CHECK-DAG: #[[$REDUCE_MAP2:.*]] = affine_map<(d0, d1) -> (d0 mod 2 + (d1 mod 2) * 2 + (d0 floordiv 2) * 8 + (d1 floordiv 2) * 4)>
+// CHECK-DAG: #[[$REDUCE_MAP1:.*]] = affine_map<(d0, d1) -> (d0 * 2 + d1 + (d1 floordiv 2) * 6)>
```
This is due to a modification I did in RAMUW. Currently whenever we call RAMUW on  memref.load with reduced_rank -- we take a separate path and call `transformMemRefLoadWithReducedRank` which unnecessary, and there is code duplication. In ``transformMemRefLoadWithReducedRank` the author does not simplify the map or compose map and operands. I've removed that special path and improved the logic in RAMUW so that all kinds of memref load/store(s) are handled. And in RAMUW we call `ullyComposeAffineMapAndOperands` and `canonicalizeMapAndOperands`, and hence we see simplified map now in the output.

bondhugula replied:
That's great - thanks.

bondhugula commented on mlir/lib/Dialect/Affine/Utils/Utils.cpp:
```diff
@@ -1338,13 +1258,25 @@ LogicalResult mlir::affine::replaceAllMemRefUsesWith(
+    // In the case of dereferencing ops not implementing
+    // AffineMapAccessInterface, we need to apply the values of `newMapOperands`
+    // to the `newMap` to get the correct indices.
+    for (unsigned i = 0; i < newMemRefRank; i++)
```
This needs braces per the new style (multi-line).

bondhugula commented on mlir/lib/Dialect/Affine/Utils/Utils.cpp:
```diff
@@ -1845,6 +1779,94 @@ LogicalResult mlir::affine::normalizeMemRef(AllocLikeOp allocOp) {
+  SmallVector<int64_t> newStaticStrides(newRank, 1);
+  SmallVector<int64_t> newStaticOffsets(newRank, 0);
+  ArrayRef<int64_t> oldShape = memrefType.getShape();
+  mlir::ValueRange oldSizes = reinterpretCastOp.getSizes();
```
You don't need `mlir::` here! Similarly, at any other places.

bondhugula commented on mlir/lib/Dialect/MemRef/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -29,6 +29,7 @@ namespace memref {
 #define DEBUG_TYPE "normalize-memrefs"
 
 using namespace mlir;
+using namespace mlir::memref;
```
Sorted order.

bondhugula commented on mlir/lib/Dialect/MemRef/Transforms/NormalizeMemRefs.cpp:
```diff
@@ -346,22 +347,31 @@ void NormalizeMemRefs::updateFunctionSignature(func::FuncOp funcOp,
+  funcOp.walk([&](Operation *op) {
+    if (auto allocOp = dyn_cast<AllocOp>(op))
+      allocOps.push_back(allocOp);
+    if (auto allocaOp = dyn_cast<AllocaOp>(op))
+      allocaOps.push_back(allocaOp);
+    if (auto reinterpretCastOp = dyn_cast<ReinterpretCastOp>(op))
```
These should all be `else if`.

bondhugula reviewed (approved):
LGTM - thanks.

arnab-polymage committed:
Normalize reinterpret_cast op

Rewrites the memref defined by reinterpet_cast op to have an identity layout map
and updates all its indexing uses. Also extend `replaceAllMemRefUsesWith` utility
to work when there are multiple occurences of `oldMemRef` in `op`'s operand list
when op is non-dereferencing.
